### PR TITLE
feat(core): pruning client by prune command

### LINKED ISSUE #1392: Pruning Store
A Full Node can convert to a Pruned Node by pruning the store and removing old blocks. Pruning the store is a time and resource consuming process and should not be performed while the node is running. We need to add a new command named prune to pactus-daemon to prune an offline node. A public method Prune should be defined in the store. This method iterates over all blocks from the pruning height to the genesis block and prunes them. The pruning height is LastBlockHeight - RetentionBlocks.

https://pips.pactus.org/PIPs/pip-27#pruning-store

### PR DESCRIPTION
**Pull request by Ja7ad**
## Description

The prune command optimizes blockchain storage by removing outdated blocks and transactions, freeing up disk space and enhancing client performance.

## Related issue(s)

- Fixes #1392

### CODE REVIEW

**Comment by Ja7ad** [comment]
@b00f I don't have idea to fix this lint error:

```shell
cmd/daemon/prune.go:99:2: use of `fmt.Printf` forbidden by pattern `^(fmt\.Print(|f|ln)|print|println)$` (forbidigo)
        fmt.Printf("\r [%s] %.0f%% Pruned: %d | Skipped: %d | Left Block: %d",
```

File: store/store.go [hunk 1]
```diff
@@ -379,25 +388,66 @@ func (s *store) IsBanned(addr crypto.Address) bool {
...
+			continue
+		}
+
+		resultFunc(0, 1, i)
+	}
+
+	return nil
+}
+
 func (s *store) pruneBlock(blockHeight uint32) (bool, error) { //nolint
```
Comment (kehiy) [inline store/store.go:426]: You can remove `//nolint` from here.

File: store/store.go [hunk 2]
```diff
@@ -379,25 +388,66 @@ func (s *store) IsBanned(addr crypto.Address) bool {
...
+	pruningHeight := cert.Height() - retentionBlocks
+
+	// start pruning blocks and transactions
+	for i := pruningHeight; i >= 1; i-- {
+		deleted, err := s.pruneBlock(i)
+		if err != nil {
+			return err
+		}
+
+		if deleted {
```
Comment (kehiy) [inline store/store.go:414]: If `deleted` is false we we had an error and we returned the function and this case won't be executed IMO.

File: store/config.go [hunk 3]
```diff
@@ -54,5 +56,11 @@ func (conf *Config) BasicCheck() error {
 		}
 	}
 
+	if conf.RetentionDays < 10 {
+		return ConfigError{
+			Reason: "retention days too small than 10 days",
```
Comment (kehiy) [inline store/config.go:61]:
```suggestion
			Reason: "Retention days can't be less than 10 days",
```

> Also I think we can set 6 days as minimum and 10 as recomended.

File: cmd/daemon/prune.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+			return
+		}
+
+		conf, _, err := cmd.MakeConfig(workingDir)
+		cmd.FatalErrorCheck(err)
+
+		cmd.PrintLine()
+		cmd.PrintWarnMsgf("Warning: This command removes all your blocks and transactions and changes " +
+			"your node to a prune node. You cannot revert to a full node.")
```
Comment (kehiy) [inline cmd/daemon/prune.go:43]: Bassicly they can revert to a full node. we can't add this to warning I think.
They can simply download the full data and start again, or use a backup.

File: cmd/daemon/prune.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		cmd.PrintLine()
+		confirmed := cmd.PromptConfirm("Do you want to continue")
```
Comment (kehiy) [inline cmd/daemon/prune.go:45]:
```suggestion
		confirmed := cmd.PromptConfirm("Do you want to continue?")
```

**Review by kehiy — CHANGES_REQUESTED** [review]

File: store/store.go [hunk 6]
```diff
@@ -379,25 +388,66 @@ func (s *store) IsBanned(addr crypto.Address) bool {
...
+
+	retentionBlocks := uint32(s.config.RetentionDays * totalBlockInDay)
+
+	if cert.Height() < retentionBlocks {
+		return nil
+	}
+
+	pruningHeight := cert.Height() - retentionBlocks
+
+	// start pruning blocks and transactions
```
Comment (kehiy) [inline store/store.go:407]: And the last thing, the logic of this loop is obvious and self explain, I think we can remove this comment.

File: store/store.go (hunk 2 above)
Comment (Ja7ad) [reply store/store.go:414]: This skipped removed or not found block.

File: store/store.go (hunk 1 above)
Comment (Ja7ad) [reply store/store.go:426]: This function needs to unit test.

Comment (kehiy) [reply store/store.go:426]: I know, I was talking about removing the comment...

File: store/store.go (hunk 2 above)
Comment (kehiy) [reply store/store.go:414]: But when its false we have an error and we won't continue, how this block get executed at all?

Comment (Ja7ad) [reply store/store.go:414]: > But when its false we have an error and we won't continue, how this block get executed at all?

- false + nil = skipped non exists block
- true + nil = block pruned
- false + err = error and stop process

File: cmd/daemon/prune.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	pruneCmd := &cobra.Command{
+		Use:   "prune",
+		Short: "prune old blocks and transactions from client",
+		Long: "The prune command optimizes blockchain storage by removing outdated blocks and transactions, " +
+			"freeing up disk space and enhancing client performance. customize pruning criteria via flags to " +
+			"manage storage effectively.",
+	}
+	parentCmd.AddCommand(pruneCmd)
+
+	workingDirOpt := pruneCmd.Flags().StringP("working-dir", "w", cmd.PactusDefaultHomeDir(),
```
Comment (b00f) [inline cmd/daemon/prune.go:27]: Can you make a function for the `working-dir` flag like:
```
func addWorkingDirOption(c *cobra.Command) *string {
	return c.Flags().StringP("working-dir", "w", cmd.PactusDefaultHomeDir(),
		"the path to the working directory that keeps the wallets and node files")
}

```

File: cmd/daemon/prune.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+			cmd.PrintErrorMsgf("The working directory is not exists: %s", workingDir)
+
+			return
+		}
+
+		conf, _, err := cmd.MakeConfig(workingDir)
+		cmd.FatalErrorCheck(err)
+
+		cmd.PrintLine()
+		cmd.PrintWarnMsgf("Warning: This command removes all your blocks and transactions and changes " +
```
Comment (b00f) [inline cmd/daemon/prune.go:56]: Can we change to this:

```
    cmd.PrintWarnMsgf("This command removes all the blocks and transactions up to %d days ago " +
    "and converts the node to prune mode.", conf.store.RetentionDays)
```

File: cmd/daemon/prune.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+			pruningProgressBar(prunedCount, skippedCount, totalCount)
+		})
+		cmd.PrintLine()
+		cmd.FatalErrorCheck(err)
+
+		str.Close()
+		_ = fileLock.Unlock()
+
+		cmd.PrintLine()
+		cmd.PrintInfoMsgf("✅ Your node successfully pruned and changed to prune client.")
```
Comment (b00f) [inline cmd/daemon/prune.go:98]:
```suggestion
		cmd.PrintInfoMsgf("✅ Your node successfully pruned and changed to prune mode.")
```

File: store/config.go [hunk 10]
```diff
@@ -54,5 +56,11 @@ func (conf *Config) BasicCheck() error {
 		}
 	}
```
Comment (b00f) [inline store/config.go:58]: Add  a function for `RetentionBlocks`
```
func (conf *NodeConfig) RetentionBlocks int {
   return conf.RetentionDays * 8640
}
```

File: store/store.go [hunk 11]
```diff
@@ -379,25 +388,65 @@ func (s *store) IsBanned(addr crypto.Address) bool {
...
 
+func (s *store) Prune(resultFunc func(pruned, skipped, pruningHeight uint32)) error {
+	cert := s.lastCertificate()
+
+	// at genesis block
+	if cert == nil {
+		return nil
+	}
+
+	retentionBlocks := uint32(s.config.RetentionDays * totalBlockInDay)
```
Comment (b00f) [inline store/store.go:399]: Use config function.

**Review by b00f — CHANGES_REQUESTED** [review]

**Review by b00f — APPROVED** [review]

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@Ja7ad Thanks.

**Pull request merged**